### PR TITLE
Incorrect behavior of "cursor: auto" over links

### PR DESCRIPTION
**Pull request by vitorroriz**
#### b5d767e418f3d45709b214a545d9a04076954823
<pre>
Incorrect behavior of &quot;cursor: auto&quot; over links
<a href="https://bugs.webkit.org/show_bug.cgi?id=173909">https://bugs.webkit.org/show_bug.cgi?id=173909</a>
rdar://99495893

Reviewed by NOBODY (OOPS!).

The initial attempt to fix this bug and comply to spec was made
at <a href="https://commits.webkit.org/253685@main">https://commits.webkit.org/253685@main</a>

After that, a:any-link no longer depends on &apos;cursor: auto&apos;
for achieving &quot;Hand&quot; (css: pointer) cursor. This is because, according to spec,
&apos;cursor: auto&apos; can only result in default or text cursor. Therefore, 253685@main
sets &apos;cursor: pointer&apos; in the UA stylesheet (html.css).

Because of that, a test case for cursor types for recognized text (live-text)
was broken. There, we were testing a &lt;img&gt; nested in a &lt;a href=&quot;&quot;&gt;. Since,
this &lt;a&gt; is now selected to have &apos;cursor: pointer&apos;, its nested &lt;img&gt;
has no longer &apos;cursor: auto&apos;, which is required for resulting in IBeam (text)
cursor for the overlay installed in the img (see EventHandler::selectCursor).

Therefore, we are updating the test case here. A &lt;img&gt; inside a &lt;a href=&quot;...&quot;&gt;
is by default styled with cursor: pointer, and therefore hovering over
its overlay doesn&apos;t show IBeam (text) cursor, unless user styles its cursor to auto.

* LayoutTests/fast/images/text-recognition/mac/cursor-types-for-recognized-text-expected.txt:
* LayoutTests/fast/images/text-recognition/mac/cursor-types-for-recognized-text.html:
Updating tests. Note that one of the test cases is commented out. As stated in the comment,
this would fail due to <a href="https://bugs.webkit.org/show_bug.cgi?id=244944">https://bugs.webkit.org/show_bug.cgi?id=244944</a> (radar: 99495849).
In short, EventHandler::selectCursor will select an IBeam cursor although it shouldn&apos;t.
This is because Node::canStartSelection() is not false for user-select: none.
* LayoutTests/platform/mac-wk2/TestExpectations:
Reseting TextExpectations for this test, before set as failure.
* Source/WebCore/html/shadow/imageOverlay.css:
(div#image-overlay):
We should not style cursor for the live-text overlay that is installed
on top of images with text. Doing that would disable user to control which
cursor is displayed for the text portion of the image, since the overlay
is in the shadow tree, hosted by the image element. I.e.: user&apos;s cursor style
for the image would not be propagated to the overlay.
</pre>

### CODE REVIEW

**Comment by hortont424** [comment]
I think that you should remove the cursor-determining code for all cases covered here from EventHandler in this patch, so that tests will reveal if you've done it right (right now you don't really know if you have, or if the native code is kicking in and sorting it out for you).

**Comment by hortont424** [comment]
Also, your changelog formatting is a bit off. There should be a bugzilla link below the title, possibly a radar below that, then a "Reviewed by NOBODY (OOPS!)." and then a short description, and then a line-by-line accounting of what you've changed, with filenames.

Did you use the tools to upload this patch? They usually build that for you.

**Review by smfr — APPROVED** [review]

**Comment by vitorroriz** [comment]
> I think that you should remove the cursor-determining code for all cases covered here from EventHandler in this patch, so that tests will reveal if you've done it right (right now you don't really know if you have, or if the native code is kicking in and sorting it out for you).

I've tried to remove but I found a problem with handling map/area elements. For a test [like this one in WPT](https://wpt.fyi/results/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/mouse-cursor-imagemap.html?label=master&label=experimental&aligned&view=subtest), I was hovering the `` being mapped and was always falling into the `auto` case of `EventHandler::selectCursor(const HitTestResult& result, bool shiftKey)`. This is because It seems that <area> doesn’t have an associated render, thus we can’t get style from it. For those cases we consider cursor to be 'auto' and rely on the auto case to generate the "link" cursor, which goes against spec. I'll turn it back to a draft and investigate more.

File: Source/WebCore/page/EventHandler.cpp [hunk 1]
```diff
@@ -1459,7 +1459,7 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
         return std::nullopt;
 
     auto renderer = node->renderer();
-    auto* style = renderer ? &renderer->style() : nullptr;
+    auto* style = renderer ? &renderer->style() : node->computedStyle();
```
Comment (nt1m) [inline Source/WebCore/page/EventHandler.cpp]: auto* style = node->computedStyle();

Also worth a comment describing why the computed style is used and not the renderStyle like it usually is the case.

Comment (vitorroriz) [reply Source/WebCore/page/EventHandler.cpp]: Can we always use the computed style here, not only when the node has no renderer?

Comment (nt1m) [reply Source/WebCore/page/EventHandler.cpp]: I think so, @anttijk wdyt?

File: Source/WebCore/page/EventHandler.cpp [hunk 2]
```diff
@@ -1459,7 +1459,7 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
         return std::nullopt;
 
     auto renderer = node->renderer();
```
Comment (nt1m) [inline Source/WebCore/page/EventHandler.cpp]: Can you move this variable further down closer to where it's used? (also it should be `auto*`)

File: Source/WebCore/page/EventHandler.cpp [hunk 3]
```diff
@@ -1459,7 +1459,7 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
...
-    auto* style = renderer ? &renderer->style() : nullptr;
+    auto* style = node->computedStyle();
```
Comment (nt1m) [inline Source/WebCore/page/EventHandler.cpp:1462]: Can you add a comment explaining why `computedStyle()` is appropriate here?

File: Source/WebCore/page/EventHandler.cpp (hunk 1 above)
Comment (vitorroriz) [reply Source/WebCore/page/EventHandler.cpp]: He DM me that it "should be ok"
